// @sentry/vercel-edge/esm/index.js
import { Integrations, RequestData } from '@sentry/core';
export { Hub, SDK_VERSION, Scope, addBreadcrumb, addEventProcessor, addGlobalEventProcessor, addIntegration, captureCheckIn, captureEvent, captureException, captureMessage, close, configureScope, continueTrace, createTransport, extractTraceparentData, flush, getActiveSpan, getActiveTransaction, getClient, getCurrentHub, getCurrentScope, getGlobalScope, getHubFromCarrier, getIsolationScope, lastEventId, makeMain, metrics, runWithAsyncContext, setContext, setExtra, setExtras, setMeasurement, setTag, setTags, setUser, spanStatusfromHttpCode, startInactiveSpan, startSpan, startSpanManual, startTransaction, trace, withMonitor, withScope } from '@sentry/core';
export { VercelEdgeClient } from './client.js';
export { defaultIntegrations, init } from './sdk.js';
import { WinterCGFetch } from './integrations/wintercg-fetch.js';

const INTEGRATIONS = {
  ...Integrations,
  WinterCGFetch,
  RequestData,
};

export { INTEGRATIONS as Integrations };
