// @sentry/vercel-edge/cjs/integrations/wintercg-fetch.js
Object.defineProperty(exports, '__esModule', { value: true });

const tracing = require('@sentry-internal/tracing');
const core = require('@sentry/core');
const utils = require('@sentry/utils');

/**
 * Creates spans and attaches tracing headers to fetch requests on WinterCG runtimes.
 */
class WinterCGFetch  {
  /**
   * @inheritDoc
   */
   static __initStatic() {this.id = 'WinterCGFetch';}

  /**
   * @inheritDoc
   */
   __init() {this.name = WinterCGFetch.id;}

    __init2() {this._createSpanUrlMap = new utils.LRUMap(100);}
    __init3() {this._headersUrlMap = new utils.LRUMap(100);}

   constructor(_options = {}) {WinterCGFetch.prototype.__init.call(this);WinterCGFetch.prototype.__init2.call(this);WinterCGFetch.prototype.__init3.call(this);
    this._options = {
      breadcrumbs: _options.breadcrumbs === undefined ? true : _options.breadcrumbs,
      shouldCreateSpanForRequest: _options.shouldCreateSpanForRequest,
    };
  }

  /**
   * @inheritDoc
   */
   setupOnce() {
    const spans = {};

    utils.addFetchInstrumentationHandler(handlerData => {
      const hub = core.getCurrentHub();
      if (!hub.getIntegration(WinterCGFetch)) {
        return;
      }

      if (core.isSentryRequestUrl(handlerData.fetchData.url, core.getClient())) {
        return;
      }

      tracing.instrumentFetchRequest(
        handlerData,
        this._shouldCreateSpan.bind(this),
        this._shouldAttachTraceData.bind(this),
        spans,
        'auto.http.wintercg_fetch',
      );

      if (this._options.breadcrumbs) {
        createBreadcrumb(handlerData);
      }
    });
  }

  /** Decides whether to attach trace data to the outgoing fetch request */
   _shouldAttachTraceData(url) {
    const client = core.getClient();

    if (!client) {
      return false;
    }

    const clientOptions = client.getOptions();

    if (clientOptions.tracePropagationTargets === undefined) {
      return true;
    }

    const cachedDecision = this._headersUrlMap.get(url);
    if (cachedDecision !== undefined) {
      return cachedDecision;
    }

    const decision = utils.stringMatchesSomePattern(url, clientOptions.tracePropagationTargets);
    this._headersUrlMap.set(url, decision);
    return decision;
  }

  /** Helper that wraps shouldCreateSpanForRequest option */
   _shouldCreateSpan(url) {
    if (this._options.shouldCreateSpanForRequest === undefined) {
      return true;
    }

    const cachedDecision = this._createSpanUrlMap.get(url);
    if (cachedDecision !== undefined) {
      return cachedDecision;
    }

    const decision = this._options.shouldCreateSpanForRequest(url);
    this._createSpanUrlMap.set(url, decision);
    return decision;
  }
} WinterCGFetch.__initStatic();

function createBreadcrumb(handlerData) {
  const { startTimestamp, endTimestamp } = handlerData;

  // We only capture complete fetch requests
  if (!endTimestamp) {
    return;
  }

  if (handlerData.error) {
    const data = handlerData.fetchData;
    const hint = {
      data: handlerData.error,
      input: handlerData.args,
      startTimestamp,
      endTimestamp,
    };

    core.addBreadcrumb(
      {
        category: 'fetch',
        data,
        level: 'error',
        type: 'http',
      },
      hint,
    );
  } else {
    const data = {
      ...handlerData.fetchData,
      status_code: handlerData.response && handlerData.response.status,
    };
    const hint = {
      input: handlerData.args,
      response: handlerData.response,
      startTimestamp,
      endTimestamp,
    };
    core.addBreadcrumb(
      {
        category: 'fetch',
        data,
        type: 'http',
      },
      hint,
    );
  }
}

exports.WinterCGFetch = WinterCGFetch;
